core/src/main/java/io/questdb/cairo/pool/ReaderPool.java

```diff
@@ -317,12 +316,7 @@ private boolean returnToPool(R reader) {
 
             e.releaseTimes[index] = clock.getTicks();
             Unsafe.arrayPutOrdered(e.allocations, index, UNALLOCATED);
-
-            // todo: there is a race condition between this method and
-            //   releaseAll() when the latter shuts down the pool. I thought of adding a version counter
-            //   that each method will attempt to increment thus recognising race condition and
-            //   taking action
-            return true;
+            return !isClosed();
```

Comment:
There is another failure that may be seen in `testGetAndCloseRace` after this change:
```
java.lang.IllegalStateException: fd 19 is already closed!
	at io.questdb/io.questdb.std.Files.auditClose(Files.java:61)
	at io.questdb/io.questdb.std.Files.close(Files.java:94)
	at io.questdb/io.questdb.std.FilesFacadeImpl.close(FilesFacadeImpl.java:44)
	at io.questdb/io.questdb.cairo.vm.MemoryCMRImpl.close(MemoryCMRImpl.java:56)
	at io.questdb/io.questdb.std.Misc.free(Misc.java:45)
	at io.questdb/io.questdb.cairo.TableReaderMetadata.close(TableReaderMetadata.java:187)
	at io.questdb/io.questdb.std.Misc.free(Misc.java:45)
	at io.questdb/io.questdb.cairo.TableReader.close(TableReader.java:163)
	at io.questdb/io.questdb.cairo.pool.ReaderPool$R.close(ReaderPool.java:370)
	at io.questdb/io.questdb.cairo.pool.ReaderPool.closeReader(ReaderPool.java:271)
	at io.questdb/io.questdb.cairo.pool.ReaderPool.releaseAll(ReaderPool.java:240)
	at io.questdb/io.questdb.cairo.pool.AbstractPool.closePool(AbstractPool.java:82)
	at io.questdb/io.questdb.cairo.pool.ReaderPool.closePool(ReaderPool.java:217)
	at io.questdb/io.questdb.cairo.pool.AbstractPool.close(AbstractPool.java:57)
	at io.questdb/io.questdb.cairo.pool.ReaderPoolTest.lambda$testGetAndCloseRace$12(ReaderPoolTest.java:390)
```

To reproduce it, you may increase the loop limit in `testGetAndCloseRace` from 100 to 1000 and then use "Run until failure" run configuration parameter in IDEA. It takes around 300-400 runs on my machine to reproduce the failure.

Reply:
Thanks Andrey, this makes sense. I know what this is - similar behaviour as we observed on WriterPool

Reply:
should be fixed in https://github.com/questdb/questdb/pull/1746/commits/47b198c1f4ad612ce29ba95a6e7e3948836d79bf

Reply:
Yes, the race seems to be fixed now. Thanks.